File: code_challenge_solution/javascript/120 Mastermind.js
let secretCode = [];
let guessCount = 0;

function generateSecretCode() {
  for (let i = 0; i < 4; i++) {
    secretCode.push(Math.floor(Math.random() * 6) + 1);
  }
}

function checkGuess(guess) {
  let blackPegs = 0;
  let whitePegs = 0;
  let codeCopy = secretCode.slice();
  let guessCopy = guess.slice();

  // Check for black pegs
  for (let i = 0; i < 4; i++) {
    if (guessCopy[i] === codeCopy[i]) {
      blackPegs++;
      guessCopy[i] = 0;
      codeCopy[i] = 0;
    }
  }

  // Check for white pegs
  for (let i = 0; i < 4; i++) {
    if (guessCopy[i] !== 0) {
      let index = codeCopy.indexOf(guessCopy[i]);
      if (index > -1) {
        whitePegs++;
        codeCopy[index] = 0;
      }
    }
  }

  return [blackPegs, whitePegs];
}

generateSecretCode();

while (guessCount < 10) {
  let guess = prompt("Enter a 4-digit guess (each digit between 1 and 6): ");
  guess = guess.split("").map(Number);

  if (guess.length !== 4) {
    alert("Invalid guess! Please enter exactly 4 digits.");
    continue;
  }

  let result = checkGuess(guess);

  console.log("Black pegs: " + result[0]);
  console.log("White pegs: " + result[1]);

  if (result[0] === 4) {
    console.log("Congratulations! You've cracked the code.");
    break;
  }

  guessCount++;
}

if (guessCount === 10) {
  console.log("Game over! You couldn't crack the code. The secret code was: " + secretCode.join(""));
}

File: code_challenge_solution/python/120 Mastermind.py
import random

# Function to generate secret code
def generate_code():
    code = []
    for _ in range(4):
        code.append(random.randint(1, 6))
    return code

# Function to get user guess
def get_guess():
    guess = []
    for _ in range(4):
        guess.append(int(input("Enter a number between 1 and 6: ")))
    return guess

# Function to evaluate guess and provide feedback
def evaluate_guess(code, guess):
    feedback = []
    for i in range(len(code)):
        if guess[i] == code[i]:
            feedback.append("red")
        elif guess[i] in code:
            feedback.append("white")
        else:
            feedback.append("black")
    return feedback

# Main game function
def play_game():
    code = generate_code()
    guess_count = 0
    while True:
        guess_count += 1
        guess = get_guess()
        feedback = evaluate_guess(code, guess)
        print("Feedback:", feedback)
        if guess == code:
            print("Congratulations! You guessed the code in", guess_count, "attempts.")
            break
        
# Start the game
play_game()